### ISSUE TITLE
colored output on linux terminal

### ISSUE THREAD

**Issue by xificurC** [factor/factor#1747]
```factor
USING: io.styles colors.constants hashtables ;
COLORS: green foreground associate "HI\n" format
```

This prints non-colored on my linux terminal. I get colored output from other applications all the time (e.g. git) so the issue shouldn't be with my terminal.

The same code outputs green text in the factor UI listener.

**Comment by mrjbq7** [comment]
The terminal doesn't support colored output using formatted streams by default (you can still get it by directly printing ANSI color codes).

However, I implemented support for 256color, so you can:

**Comment by mrjbq7** [comment]
https://github.com/factor/factor/blob/master/extra/io/streams/256color/256color.factor

Wrap your printing in `[ ... ] with-256color`.

Or start a child listener like this and it will support natively:

`[ listener ] with-256color`

If your terminal doesn't support 256 colors and you just want he basic ANSI colors, you can use `io.streams.ansi`.

**Comment by xificurC** [comment]
That was prompt, thanks! I will look into the mentioned options. Would it make sense to note this in the `io.styles` docs?

**Comment by mrjbq7** [comment]
(It's not perfect, in some cases color codes take up formatting space in things like table formatting)

And one could argue we should support it in the listener directly without requiring the user to use another vocal like these, but.... it works!

:-)

**Comment by mrjbq7** [comment]
Yes, it would be good to document somewhere. I would welcome a patch. I'm about to get on a flight, so will not be able to respond in 2 minutes.

**Comment by mrjbq7** [comment]
I'm also not sure how to support/detect light vs dark backgrounds in the terminal automatically. That was another thing preventing us making color mode the default.

**Comment by xificurC** [comment]
With `io.streams.256color` you meant I could use my first example wrapped in a quotation, right? As in

```factor
[ COLOR: green foreground associate "HI\n" format ] with-256color
```

That gives me:

```
Generic word nth does not define a method for the fixnum class.
Dispatching on object: 72
```

Same for `io.streams.ansi`

**Comment by xificurC** [comment]
Have a nice flight ;)

**Comment by AlexIljin** [comment]
You are missing a `swap` before `format`. 11.11.2016, 18:50, "Peter Nagy" <notifications@github.com>:With io.streams.256color you meant I could use my first example wrapped in a quotation, right? As in[ COLOR: green foreground associate "HI\n" format ] with-256colorThat gives me:Generic word nth does not define a method for the fixnum class.
Dispatching on object: 72
Same for io.streams.ansi ---=====---Александр

**Comment by mrjbq7** [comment]
As @AlexIljin says you need to ``swap`` since ``format`` expects ``( str style -- )``.

```factor
USING: io.styles colors.constants hashtables ;

[
    COLOR: green foreground associate "HI\n" swap format
] with-256color
```

Or, just switch to a "color-supported listener":

```factor
[ listener ] with-256color
```

Then you can just run code that expects to color output directly without using ``with-256color`` again.  Note: if you want to exit the listener, you need two ``Ctrl-D``, the first to exit the "color-supported listener" and the second to exit the original listener.

**Comment by mrjbq7** [comment]
Also, instead of using ``associate``, if you have non-dynamic format strings, you can just use:

```factor
"Hello" H{ { foreground COLOR: green } } format
```

**Comment by xificurC** [comment]
Duh, don't I feel stupid. The corrected examples work as advertised.

Is there any ongoing work with the error messages? E.g. in this example the message has little correlation to the toplevel code.

**Comment by mrjbq7** [comment]
It would be nice to catch errors like that earlier and provide nicer errors back, often we do a kind of "duck-typing" and errors happen when you try and use an object and it doesn't support that way of being used... but in common cases we can and should probably lift those errors up higher and provide a nicer message.

**Comment by AlexIljin** [comment]
John, would you agree that using TYPED: in this case would solve the "bad error message" issue, and not introduce any significant overhead? 12.11.2016, 23:09, "John Benediktsson" <notifications@github.com>:It would be nice to catch errors like that earlier and provide nicer errors back, often we do a kind of "duck-typing" and errors happen when you try and use an object and it doesn't support that way of being used... but in common cases we can and should probably lift those errors up higher and provide a nicer message.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.   ---=====---Александр

**Comment by mrjbq7** [comment]
TLDR: it's complicated.

Well, maybe it's not quite so easy in this case -- I suppose we would want to enforce that ``style`` is an ``assoc`` and maybe that ``str`` is a ``string`` although some output streams might support other objects being "formatted" (for example, the UI listener supports printing arrays-of-strings to mean "lines of text").

But we have a particular problem here that ``format`` just delegates to ``stream-format``:

```factor
: format ( str style -- ) output-stream get stream-format ;
```

And ``stream-format`` is a generic (which currently doesn't support type checking at the generic dispatch level):

```factor
GENERIC: stream-format ( str style stream -- )
```

Which would imply that we should do the type checking in the implementations, of which there are several:

```factor
jbenedik@red ~/Projects/factor $ [master*] factorgrep "M: " | grep stream-format./basis/html/streams/streams.factor:M: html-writer stream-format
./basis/io/styles/styles.factor:M: style-stream stream-format
./basis/io/styles/styles.factor:M: plain-writer stream-format
./basis/ui/gadgets/panes/panes.factor:M: pane-stream stream-format
./basis/ui/gadgets/panes/panes.factor:M: pack stream-format
./basis/ui/gadgets/panes/panes.factor:M: paragraph stream-format
./extra/io/streams/256color/256color.factor:M: 256color stream-format
./extra/io/streams/ansi/ansi.factor:M: ansi stream-format
./extra/pdf/streams/streams.factor:M: pdf-writer stream-format
```

So basically, we would need all this duplicate type-checking everywhere, which we could add... or maybe have a "lookup-style" word that everyone uses instead of ``at`` that does type checking and nice error messages.

Anyway, suggestions and patches are welcome!  The ``format`` versus ``swap format`` is an error I've made before and generally I'd love nicer error messages.
